### DISCUSSION TITLE
Generating Config Key (CK) and Browser Exam Key (BEK) with SEB Server

### DISCUSSION THREAD

**Discussion by ano010** [SafeExamBrowser/seb-server#57] [Q&A]
Hi,

I want to check whether there is an option in the SEB server to generate Browser Exam Key. I can see there is an option for generating Config Key but I couldn't find any option for generating Browser Exam Key in SEB exam configuration settings UI.

I have been trying to validate CK and BEK in Moodle using the SEB plugin for Moodle. But the Moodle only accepts if both keys are set. I was able to configure the exam and downloaded the configuration. But as there is no option for creating BEK from the SEB server. I can't test this functionality.

Thanks in Advance
I would appreciate any help.

**Comment by anhefti** [comment]
Hi, 
Since the Browser Exam Key is been generated from the SEB binary, SEB Server is not able to generate this key itself.
What you can do is just use the CK or, if you want to use both, you can download the SEB Settings of the Exam Configuration from SEB Server and load it into the SEB Config Tool of the SEB you want to use for the exam. Then you are able to generate the BEK with the SEB Config Tool.
Now you can get the CK from SEB Server and the BEK from Config Tool and set it within the Moodle configuration of the course/quiz.

Please note that the automated CK restriction for SEB Server Moodle integration is still not implements yet (like in Open edX integration for example) but is expected to coming soon with the next one or two version of SEB Server.

Best regards

**Reply by ano010** [reply]
@anhefti,
 Thanks for the prompt reply. I need to configure some additional settings for example proctoring setting. As some of the additional settings are not in SEB Config Tool, The CK generated by the SEB server and the CK generated by the SEB Config Tool is different. So BKE calculated with SEB Config Tool is not valid for the actual configuration I downloaded from the SEB server.

However, using only CK is enough solution as I am looking to validate only the settings, not versions. But I couldn't find any way to use only the CK in the SEB plugin for Moodle. Do you have any idea whether it is possible in Moodle?

Can you explain a little more about the automated CK restriction for SEB Server Moodle integration? Though it is not implemented yet. I can get some idea if it is useful in future?

**Reply by danschlet** [reply]
Btw. if you open any SEB config file in the SEB Config Tool, even if it contains settings which the config tool doesn't know, the calculated Browser Exam Key will be correct for that file and the SEB Windows version you used the SEB Configuration Tool of. This is important because you need to be able to for example calculate Windows BEK for config files containing Mac SEB settings which the config tool doesn't know (yet).

**Reply by anhefti** [reply]
@ano010 Sorry, according to the below explanation of @danschlet , my answer was partially wrong since I was not aware that CK cannot be manually set on Moodle deeper integration. Please also have a look to an older question and an workaround explained [here](https://github.com/SafeExamBrowser/seb-server/issues/25)

> Can you explain a little more about the automated CK restriction for SEB Server Moodle integration? Though it is not implemented yet. I can get some idea if it is useful in future?

We are currently working on a Moodle plugin for better SEB Server integration that also includes automated SEB restriction for Moodle. Within this feature (like in the Open edX integration) SEB Server automatically calculates the CK and send it to Moodle for restriction when the exam on SEB Server starts. If there shall be a EBK involved it is also possible to manually set/register such EBKs on the SEB Server exam and they also get pushed to Moodle by the SEB Server for restriction when the exam starts.
For further information please have a look at: https://seb-server.readthedocs.io/en/latest/exam_restriction.html

**Reply by ano010** [reply]
> Btw. if you open any SEB config file in the SEB Config Tool, even if it contains settings which the config tool doesn't know, the calculated Browser Exam Key will be correct for that file and the SEB Windows version you used the SEB Configuration Tool of. This is important because you need to be able to for example calculate Windows BEK for config files containing Mac SEB settings which the config tool doesn't know (yet).

@danschlet,

But it doesn't work for me. As per your explanation, It should be applied to the Config Key as well. right? I have retested the config key generation. Stil the CK generated by the SEB server and the CK generated by the SEB config tool is different. It seems the CK generated by the Config Tool is correct since it works when I add it in Moodle with the BEK.  But it does not work when I add the CK generated by SEB server with the BEK. In that case, can we assume any bug in creating CK in the SEB server?
 I am using SEB Server Version 1.3-statble.

**Reply by danschlet** [reply]
You probably didn't understand fully: You cannot add the CK manually to a Moodle quiz. That's only possible with the BEK and that's complicated (as you need another for any SEB version/platform and need to make sure the config file isn't altered).

**Reply by ano010** [reply]
I understand that. So I am thinking of using CK and all BEKs in Moodle. I have only one confusion.  The CK generated by the SEB server is different than the Config Tool when I load the same configuration downloaded from the SEB server. I haven't altered anything.  So can we assume there is something wrong in generating CK in SEB sever? 

Or do you think CK generated by the SEB server and Config Tool can be different? I would really appreciate a brief explanation about this.

**Reply by anhefti** [reply]
I've tried this with several configurations now on my local SEB Server. I always get the same CK when I load the downloaded SEB Settings within the Config Tool:

![image](https://user-images.githubusercontent.com/39078777/170251724-0733807c-678f-4ccc-8066-a4fdc1add21f.png)

Please make sure you always have saved/published your changes of the SEB Settings on the SEB Server before downloading the settings. If this is still an issue, it would be good if you can apply your specific configuration file, so we can have a look into it, import it to SEB Server and load it with the Config Tool.

**Reply by ano010** [reply]
Thanks for the detailed explanation. I was able to find out my mistake. I have been using the configuration downloaded from "Export Exam Connection Config" as shown in the below screenshot.

<img width="883" alt="Screenshot 2022-05-27 141452" src="https://user-images.githubusercontent.com/33832254/170668210-0c7fb3ed-620e-4d58-b3db-68f4541dc855.png">

But still, I wonder in which way does the "Exam Configuration" as it is shown in your image and "Exam Connection Configuration" for a specif exam as it is shown in my image differ?

**Reply by anhefti** [reply]
Glad to read that it seems to work correctly now. 
When you are working with the SEB Server, then there are two different configurations involved:

**The (Exam) Connection Configuration**: That only has detailed settings for SEB to connect to the SEB Server. This configuration does originally not depend on an specific exam and can be used within different scenarios that fits your need. For example you can create only one Connection Configuration and download that from the Connection Configuration page to use with SEB. You can send this with Mail or link it within a public portal for example. SEB users that loads this configuration will be connected to SEB Server and get a list of running exams from the SEB Server presented. SEB user can then choose one of the exams. Alternatively, you can download a Connection Configuration directly within the Exam to short-cut the previous step. A SEB user, after connecting with the SEB Server will then automatically be directed to the specified exam without the presentation of the running exam list.

**The Exam Configuration**: This configuration is equivalent to a SEB settings configuration for starting an exam within the SEB Config Tool. Here you can set all the SEB Settings you need for a specific Exam. An Exam Configuration can then be bound to an Exam and is automatically downloaded by a connecting SEB client that selected an Exam on the SEB Server. The Config-Key only apply to this configuration.

Please also read the [User Manual](https://seb-server.readthedocs.io/en/latest/configurations.html) for more detailed information.

**Comment by danschlet** [comment]
Currently you cannot enter a CK manually in Moodle. But you can use the Moodle quiz settings option "Yes - Upload my own config" and upload that config file where you added those additional settings. You then need to make sure that the SEB client can reconfigure to the config file which Moodle will automatically send when the quiz is started. For that you need to set the correct following SEB settings in SEB Server for that exam: 
Allow reconfiguring secure/exam session: enable
Reconfiguring URL: (URL to the root path of your Moodle instance)/* (or just the * wildcard to test it)
Clear browser session when ending an exam or terminating SEB: disable

**Reply by ano010** [reply]
@danschlet,

Few clarifications, 

1. In which files should I set the reconfiguration settings?  with the config file uploaded in Moodle using the option "Yes - Upload my own config" or with the config file used to start the exam?
2. Can I create both config files from the SEB server?
3. If we use the "Yes - Upload my own config" option and upload the correct configuration in Moodle, then there is no possibility of using a different configuration than the uploaded one to start the quiz. right? so the user can log in to Moodle, download the configuration, and start the exam using that configuration if I am right? So how do the reconfiguration settings help here?
4. If the reconfiguration needs to work, the user should not change the reconfiguration settings in the config file used to start the quiz. right?

**Reply by ano010** [reply]
Hi,

There are two types of SEB configuration I can see when creating connection configuration which is "Configure a client" and "Starting an exam". I couldn't find any use case for "Configure a client". Can you tell for what purpose it can be used?

**Comment by RahulKrishna05** [comment]


**Reply by anhefti** [reply]
Hi, the Browser Exam Key is not generated by SEB Server but only by SEB Clients. 
If you use SEB Windows you can use the SEB Configuration Tool to extract the Browser Exam Key:

<img width="1259" height="721" alt="image" src="https://github.com/user-attachments/assets/0c52843d-428c-49c1-a39f-9447f8fadd2e" />

Make sore you loaded also your configuration with the Config Tool before extracting the Key.

For SEB Mac there should be similar approach on Mac SEB Configuration Tool
